utf8/reader_test.go

```diff
@@ -0,0 +1,288 @@
+package utf8_test
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"testing"
+
+	stdiotest "testing/iotest"
+
+	"github.com/madlambda/spells/assert"
+	"github.com/madlambda/spells/io/runes"
+	"github.com/madlambda/spells/iotest"
+	"github.com/madlambda/spells/utf8"
+)
+
+const expectedString = "Εν οίδα ότι ουδέν οίδα"
+
+func TestUTF8Reader(t *testing.T) {
+	type testcase struct {
+		input  string
+		repeat int
+		err    error
+		errOut []rune
+	}
+
+	testcases := []testcase{
+		{
+			input:  "",
+			repeat: 1,
+		},
+		{
+			input:  string([]byte{0x80}),
+			repeat: 1,
+			err:    utf8.ErrDecode,
+		},
+		{
+			input:  "test",
+			repeat: 1,
+		},
+		{
+			input:  "Εν οίδα ότι ουδέν οίδα",
+			repeat: 1,
+		},
+		{
+			input:  "Εν οίδα ότι TESTE ουδέν οίδα",
+			repeat: 1,
+		},
+		{
+			input:  "test\n",
+			repeat: 1024,
+		},
+		{
+			input:  "Εν οίδα ότι ουδέν οίδα",
+			repeat: 1024,
+		},
+		{
+			input:  "Εν οίδα ότι ουδέν οίδα",
+			repeat: 1025,
+		},
+		{
+			input:  "Εν οίδα ότι ουδέν οίδα",
+			repeat: 2048,
+		},
+		{
+			input:  "Εν οίδα ότι ουδέν οίδα",
+			repeat: 2049,
+		},
+		{
+			input:  string([]byte{206, 149, 206, 206, 206}),
+			repeat: 1,
+			err:    utf8.ErrDecode,
+			errOut: []rune{'Ε'},
+		},
+	}
+
+	// test NewReaderReader
+	for _, tc := range testcases {
+		reader := utf8.NewReaderReader(iotest.NewRepeatReader(
+			bytes.NewBuffer([]byte(tc.input)),
+			tc.repeat))
+
+		got, readErr := runes.ReadAll(reader)
+		assert.IsError(t, tc.err, readErr, "Read() error")
+
+		var expected []rune
+
+		if readErr == nil {
+			repeater := iotest.NewRepeatReader(bytes.NewBuffer([]byte(tc.input)),
+				tc.repeat)
+
+			expectedBytes, err := io.ReadAll(repeater)
+			assert.NoError(t, err, "repeating expected")
+
+			expected = []rune(string(expectedBytes))
+		} else {
+			expected = tc.errOut
+		}
+
+		assert.EqualInts(t, len(expected), len(got), "rune slice len mismatch: %s",
+			string(got))
+
+		for i, r := range expected {
+			if r != got[i] {
+				t.Errorf("want[%d = %c] but got[%d = %c]", r, r, got[i], got[i])
```

Comment:
```suggestion
				t.Errorf("at [%d] want[%c] != got[%c]", i, r, got[i])
```

Reply:
no, I wanted to show the rune integer value and its character printed.